File: core/java/src/main/java/hara/lang/protocol/IIter.java
package hara.lang.protocol;

import java.util.Iterator;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** Produces a fresh traversal cursor for a value. */
@HaraProtocolBinding(namespace = "std.protocol.iiter", name = "IIter")
public interface IIter<E> {
  @HaraMethod(value = "iter", arity = 1)
  Iterator<E> iter();
}

File: core/rust/src/lang/protocol/iiter.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(namespace = "std.protocol.iiter", name = "IIter")]
pub trait IIter {
    type Item;
    type Iter: Iterator<Item = Self::Item>;

    #[hara_method(value = "iter", arity = 1)]
    fn iter(&self) -> Self::Iter;
}
